src/data/set/basic.lean

```diff
@@ -1765,16 +1752,163 @@ For the corresponding result for `subtype`, see `subtype.subsingleton`. -/
 instance subsingleton_coe_of_subsingleton [subsingleton α] {s : set α} : subsingleton s :=
 by { rw [s.subsingleton_coe], exact subsingleton_of_subsingleton }
 
+theorem subsingleton_mono {α : Type*} {s t : set α} (hst : s ⊆ t) (hs : subsingleton t) :
+  subsingleton s := (subsingleton_coe _).2 $ (t.subsingleton_coe.1 hs).anti hst
+
+/-- The image of a subsingleton is a subsingleton. -/
+lemma subsingleton.image (hs : s.subsingleton) (f : α → β) : (f '' s).subsingleton :=
+λ _ ⟨x, hx, Hx⟩ _ ⟨y, hy, Hy⟩, Hx ▸ Hy ▸ congr_arg f (hs hx hy)
+
 /-- The preimage of a subsingleton under an injective map is a subsingleton. -/
 theorem subsingleton.preimage {s : set β} (hs : s.subsingleton) {f : α → β}
-  (hf : function.injective f) :
-  (f ⁻¹' s).subsingleton :=
-λ a ha b hb, hf $ hs ha hb
+  (hf : function.injective f) : (f ⁻¹' s).subsingleton := λ a ha b hb, hf $ hs ha hb
 
-/-- `s` is a subsingleton, if its image of an injective function is. -/
+/-- If the image of a set under an injective map is a subsingleton, the set is a subsingleton. -/
 theorem subsingleton_of_image {α β : Type*} {f : α → β} (hf : function.injective f)
   (s : set α) (hs : (f '' s).subsingleton) : s.subsingleton :=
-(hs.preimage hf).mono $ subset_preimage_image _ _
+(hs.preimage hf).anti $ subset_preimage_image _ _
+
+/-- If the preimage of a set under an surjective map is a subsingleton,
+the set is a subsingleton. -/
+theorem subsingleton_of_preimage {α β : Type*} {f : α → β} (hf : function.surjective f)
+  (s : set β) (hs : (f ⁻¹' s).subsingleton) : s.subsingleton :=
+λ fx hx fy hy, by { rcases ⟨hf fx, hf fy⟩ with ⟨⟨x, rfl⟩, ⟨y, rfl⟩⟩, exact congr_arg f (hs hx hy) }
+
+/-! ### Nontrivial -/
+
+/-- A set `s` is `nontrivial` if it has at least two distinct elements. -/
+protected def nontrivial (s : set α) : Prop := ∃ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), x ≠ y
```

Comment:
Is there a reason for this specific argument order? I would have expected
```suggestion
protected def nontrivial (s : set α) : Prop := ∃ x y ∈ s, x ≠ y
```

Reply:
These are the same. You'd expect they wouldn't be (that `∃ x y ∈ s, x ≠ y` would have the order `x y hx hy`) but that isn't true. So I was simply mimicking the way `subsingleton` does it. But originally I had what you suggest, and changed it at the end with no other changes needed.

Reply:
```
lemma nontrivial_def : s.nontrivial ↔ ∃ x y ∈ s, x ≠ y := iff.rfl

lemma nontrivial_def' : s.nontrivial ↔ ∃ x y (hx : x ∈ s) (hy : y ∈ s), x ≠ y := sorry
```
See this: it is rfl to the former but not the latter.

Reply:
I wouldn't expect either, this seems like a trivial implementation detail. I'd recommend just going with what's visually simplest, i.e. my approach.

Reply:
Well, I don't disagree, I like that better also.

Reply:
Change made.